### PR TITLE
ompi_setup_fc.m4: use -Wl,-ld_classic if supported

### PR DESCRIPTION
**Pull request by jsquyres**
Per https://github.com/open-mpi/ompi/issues/12427, on MacOS, add -Wl,-ld_classic to the Fortran wrapper compiler if that flag is needed.
    
Specifically, Open MPI has used -Wl,-commons,use_dylibs for decades to support common symbols (e.g., MPI_BOTTOM) in the Fortran bindings. There is a window of Xcode versions where this switch was effectively disabled; it effectively required the additional -Wl,-ld_classic switch to force the use of the "old" Apple linker (that still supported -Wl,-commons,use_dylibs).  Update the configury to test whether we need -Wl,-ld_classic or not.

Refs #12427

FYI @jeffhammond @fxcoudert @ggouaillardet

### CODE REVIEW

**Comment by fxcoudert** [comment]
I think that, with Xcode 16 beta, the support of `-commons use_dylibs` was added to the new linker. So `-ld_classic` really should only be passed if `-commons use_dylibs` is otherwise rejected.

**Comment by jsquyres** [comment]
@fxcoudert Ah -- interesting point!  That certainly makes things more complicated.  I'll push an update here that seems to work for me -- can you check it as well?

I'm going to switch this PR back to draft status because I'm not entirely happy with how I coded up the configury; @bwbarrett can you chime in with any better suggestions on how to build the shared library in `_OMPI_SETUP_FC_XCODE_COMMONS_LDFLAGS_BACKEND`?  (the `libtool` executable won't exist yet)

**Comment by jsquyres** [comment]
The above force-push was to update/clarify some of the explanation comments -- no code or logic changed.

**Comment by jsquyres** [comment]
@bwbarrett and I talked through this on the RM call today; this PR seems to be a reasonable compromise.  Move this PR out of Draft state.

@fxcoudert and/or anyone else: can you test this on your systems and see if it fixes the issue for you?

File: config/ompi_setup_fc.m4 [hunk 1]
```diff
@@ -50,9 +50,168 @@ AC_DEFUN_ONCE([_OMPI_SETUP_FC_COMPILER],[
...
+    OPAL_VAR_SCOPE_PUSH([xcode_happy xcode_dir LDFLAGS_save_xcode LIBS_save_xcode])
+
+    xcode_dir=conftest.$$
+    rm -rf $xcode_dir
+    mkdir -p $xcode_dir
+    cd $xcode_dir
+
+    LIBS_save_xcode=$LIBS
+    LDFLAGS_save_xcode=$LDFLAGS
+    LDFLAGS="$LDFLAGS -L. $1"
```
Comment (bwbarrett) [inline config/ompi_setup_fc.m4:127]: Why are you dropping all of LDFLAGS?  We wouldn't catch conflicting flags on picking the linker version (`-ld_new`) or incompatible options (`-merge*`).

File: config/ompi_setup_fc.m4 [hunk 2]
```diff
@@ -50,9 +50,168 @@ AC_DEFUN_ONCE([_OMPI_SETUP_FC_COMPILER],[
...
+dnl * It seems to fail (by itself) when building shared libraries for
+dnl   certain versions of Xcode.  In these failure scenarios, we add
+dnl   -Wl,-ld_classic to make it work.
+dnl
+dnl Note: the -ld_classic flag works with both static and shared
+dnl libraries, so it is safe to always do this test -- even if
+dnl we are only building static libraries.
+dnl
+dnl The history is that for a long time (decades),
+dnl -Wl,-commons,use_dylibs worked by itself.  But then Apple decided
```
[the remark this answers is not in the document]
Comment (jsquyres) [reply config/ompi_setup_fc.m4:67]: Will update.

File: config/ompi_setup_fc.m4 [hunk 3]
```diff
@@ -50,9 +50,168 @@ AC_DEFUN_ONCE([_OMPI_SETUP_FC_COMPILER],[
...
+                          [xcode_happy=0])])
+    AC_LANG_POP
+
+    # Exit the temp dir
+    cd ..
+    rm -rf $xcode_dir
+
+    # LIBS was set specifically for the artificial conditions of this
+    # test, so reset it
+    LIBS=$LIBS_save_xcode
```
Comment (jsquyres) [inline config/ompi_setup_fc.m4:196]: Need to remove `-L.` from `LDFLAGS` here.

File: config/ompi_setup_fc.m4 (hunk 1 above)
Comment (jsquyres) [reply config/ompi_setup_fc.m4:127]: @bwbarrett and I talked this through IRL: all is ok.

**Comment by jsquyres** [comment]
@bwbarrett All issues have been addressed; please re-review.

**Review by bwbarrett — APPROVED** [review]

**Pull request merged**